### ISSUE TITLE
AGW: MME: memory leak report from scale test

### ISSUE THREAD

**Issue by pshelar** [magma/magma#5086]
### Your Environment

- **Version:** 1.4
- **Affected Component:** AGW
- **Affected Subcomponent:** MME
- **Deployment Environment:** bare-metal (AGW)

### Describe the Issue
Saw memory leak report in syslog:
```
Feb 22 16:08:53 agw1 magmad[25148]: task: <Task pending coro=<Job._sleep_for_interval() running at /usr/local/lib/python3.5/dist-packages/magma/common/job.py:100>>
Feb 22 16:08:55 agw1 mme[26668]: Cleaning up SPGW configs
Feb 22 16:08:55 agw1 mme[26668]: =================================================================
Feb 22 16:08:55 agw1 mme[26668]: ==26668==ERROR: LeakSanitizer: detected memory leaks
Feb 22 16:08:55 agw1 mme[26668]: Direct leak of 160 byte(s) in 2 object(s) allocated from:
Feb 22 16:08:55 agw1 mme[26668]:     #0 0x7f56e2ff1cdf in operator new[](unsigned long) (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xdcdf)
Feb 22 16:08:55 agw1 mme[26668]:     #1 0x7f56e01b4783 in fluid_base::BaseOFConnection::OFReadBuffer::read_notify(unsigned short) fluid/base/BaseOFConnection.cc:84
Feb 22 16:08:55 agw1 mme[26668]:     #2 0x7f56e01b4783 in fluid_base::BaseOFConnection::LibEventBaseOFConnection::read_cb(bufferevent*, void*) fluid/base/BaseOFConnection.cc:353
Feb 22 16:08:55 agw1 mme[26668]: Direct leak of 160 byte(s) in 10 object(s) allocated from:
Feb 22 16:08:55 agw1 mme[26668]:     #0 0x7f56e2ff114e in __interceptor_calloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xd14e)
Feb 22 16:08:55 agw1 mme[26668]:     #1 0x5591e64e145f in _create_new_attach_info /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/Attach.c:2508
Feb 22 16:08:55 agw1 mme[26668]:     #2 0x5591e64e3e73 in emm_proc_attach_request /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/Attach.c:484
Feb 22 16:08:55 agw1 mme[26668]:     #3 0x5591e650546e in emm_recv_attach_request /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_recv.c:374
Feb 22 16:08:55 agw1 mme[26668]:     #4 0x5591e6501067 in _emm_as_recv /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_as.c:377
Feb 22 16:08:55 agw1 mme[26668]:     #5 0x5591e6503a43 in _emm_as_data_ind /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_as.c:678
Feb 22 16:08:55 agw1 mme[26668]:     #6 0x5591e6503a43 in emm_as_send /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_as.c:180
Feb 22 16:08:55 agw1 mme[26668]:     #7 0x5591e64f3596 in emm_sap_send /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_sap.c:105
Feb 22 16:08:55 agw1 mme[26668]:     #8 0x5591e64d5d71 in nas_proc_ul_transfer_ind /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/nas_proc.c:326
Feb 22 16:08:55 agw1 mme[26668]:     #9 0x5591e6213022 in handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/mme_app/mme_app_main.c:98
Feb 22 16:08:55 agw1 mme[26668]:     #10 0x7f56e03e37bd in zloop_start (/usr/lib/x86_64-linux-gnu/libczmq.so.4+0x287bd)
Feb 22 16:08:55 agw1 mme[26668]: Direct leak of 96 byte(s) in 6 object(s) allocated from:
Feb 22 16:08:55 agw1 mme[26668]:     #0 0x7f56e2ff0c6f in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xcc6f)
Feb 22 16:08:55 agw1 mme[26668]:     #1 0x5591e618d75e in blk2bstr /home/vagrant/magma/lte/gateway/c/oai/lib/bstr/bstrlib.c:314
Feb 22 16:08:55 agw1 mme[26668]:     #2 0x5591e64f11f3 in _emm_cn_cs_response_success /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_cn.c:588
Feb 22 16:08:55 agw1 mme[26668]:     #3 0x5591e64f245a in emm_cn_send /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_cn.c:1154
Feb 22 16:08:55 agw1 mme[26668]:     #4 0x5591e64f356c in emm_sap_send /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_sap.c:113
Feb 22 16:08:55 agw1 mme[26668]:     #5 0x5591e64d6608 in nas_proc_cs_respose_success /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/nas_proc.c:485
Feb 22 16:08:55 agw1 mme[26668]:     #6 0x5591e62155ff in mme_app_handle_create_sess_resp /home/vagrant/magma/lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:1248
Feb 22 16:08:55 agw1 mme[26668]:     #7 0x5591e6213052 in handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/mme_app/mme_app_main.c:116
Feb 22 16:08:55 agw1 mme[26668]:     #8 0x7f56e03e37bd in zloop_start (/usr/lib/x86_64-linux-gnu/libczmq.so.4+0x287bd)
Feb 22 16:08:55 agw1 mme[26668]: Direct leak of 64 byte(s) in 4 object(s) allocated from:
Feb 22 16:08:55 agw1 mme[26668]:     #0 0x7f56e2ff0c6f in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xcc6f)
Feb 22 16:08:55 agw1 mme[26668]:     #1 0x5591e618d75e in blk2bstr /home/vagrant/magma/lte/gateway/c/oai/lib/bstr/bstrlib.c:314
Feb 22 16:08:55 agw1 mme[26668]:     #2 0x5591e61de799 in s1ap_mme_handle_initial_context_setup_response /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:975
Feb 22 16:08:55 agw1 mme[26668]:     #3 0x5591e61e69f5 in s1ap_mme_handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:205
Feb 22 16:08:55 agw1 mme[26668]:     #4 0x5591e619741f in handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:141
Feb 22 16:08:55 agw1 mme[26668]:     #5 0x7f56e03e37bd in zloop_start (/usr/lib/x86_64-linux-gnu/libczmq.so.4+0x287bd)
Feb 22 16:08:55 agw1 mme[26668]: Indirect leak of 192 byte(s) in 6 object(s) allocated from:
Feb 22 16:08:55 agw1 mme[26668]:     #0 0x7f56e2ff0c6f in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xcc6f)
Feb 22 16:08:55 agw1 mme[26668]:     #1 0x5591e618d781 in blk2bstr /home/vagrant/magma/lte/gateway/c/oai/lib/bstr/bstrlib.c:323
Feb 22 16:08:55 agw1 mme[26668]:     #2 0x5591e64f11f3 in _emm_cn_cs_response_success /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_cn.c:588
Feb 22 16:08:55 agw1 mme[26668]:     #3 0x5591e64f245a in emm_cn_send /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_cn.c:1154
Feb 22 16:08:55 agw1 mme[26668]:     #4 0x5591e64f356c in emm_sap_send /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/emm/sap/emm_sap.c:113
Feb 22 16:08:55 agw1 mme[26668]:     #5 0x5591e64d6608 in nas_proc_cs_respose_success /home/vagrant/magma/lte/gateway/c/oai/tasks/nas/nas_proc.c:485
Feb 22 16:08:55 agw1 mme[26668]:     #6 0x5591e62155ff in mme_app_handle_create_sess_resp /home/vagrant/magma/lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:1248
Feb 22 16:08:55 agw1 mme[26668]:     #7 0x5591e6213052 in handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/mme_app/mme_app_main.c:116
Feb 22 16:08:55 agw1 mme[26668]:     #8 0x7f56e03e37bd in zloop_start (/usr/lib/x86_64-linux-gnu/libczmq.so.4+0x287bd)
Feb 22 16:08:55 agw1 mme[26668]: Indirect leak of 32 byte(s) in 4 object(s) allocated from:
Feb 22 16:08:55 agw1 mme[26668]:     #0 0x7f56e2ff0c6f in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xcc6f)
Feb 22 16:08:55 agw1 mme[26668]:     #1 0x5591e618d781 in blk2bstr /home/vagrant/magma/lte/gateway/c/oai/lib/bstr/bstrlib.c:323
Feb 22 16:08:55 agw1 mme[26668]:     #2 0x5591e61de799 in s1ap_mme_handle_initial_context_setup_response /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:975
Feb 22 16:08:55 agw1 mme[26668]:     #3 0x5591e61e69f5 in s1ap_mme_handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:205
Feb 22 16:08:55 agw1 mme[26668]:     #4 0x5591e619741f in handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:141
Feb 22 16:08:55 agw1 mme[26668]:     #5 0x7f56e03e37bd in zloop_start (/usr/lib/x86_64-linux-gnu/libczmq.so.4+0x287bd)
Feb 22 16:08:55 agw1 mme[26668]: SUMMARY: LeakSanitizer: 704 byte(s) leaked in 32 allocation(s).
```

**Comment by ulaskozat** [comment]
@pshelar what is the context here? Was mme restarted at the end of the test, was there a crash, etc.?

**Comment by ardzoht** [comment]
All of these leaks have been fixed by according mem leaks fixes / are not reproducible either on v1.5 or master.
